Subject: Convert build system to automake
Origin: Unknown
Forwarded: no
Acked-by: Jan Engelhardt <jengelh@computergmbh.de>
Last-Update: 2010-04-11
--- /dev/null
+++ b/Makefile.am
@@ -0,0 +1,24 @@
+AUTOMAKE_OPTIONS   = foreign subdir-objects
+moduledir          = @PAM_MODDIR@
+module_LTLIBRARIES = pam_shield.la
+sbin_PROGRAMS      = shield-purge
+#sbin_SCRIPTS       = shield-trigger.sh shield-trigger-iptables.sh
+securitydir        = ${sysconfdir}/security
+security_DATA      = shield.conf
+dist_man_MANS = man/shield-purge.8.gz man/shield-trigger.8.gz man/shield-trigger-iptables.8.gz
+
+AM_CFLAGS = -D_LARGEFILE_SOURCE=1 -D_LARGE_FILES -D_FILE_OFFSET_BITS=64 \
+            -D_REENTRANT -Wall -Wmissing-declarations -Wmissing-prototypes \
+            -Wredundant-decls -Wshadow -Wstrict-prototypes -Winline -pipe
+
+pam_shield_la_LDFLAGS = -module -avoid-version
+pam_shield_la_LIBADD  = -lgdbm -lpam
+pam_shield_la_SOURCES = pam_shield.c pam_shield_lib.c
+shield_purge_LDFLAGS  = -lgdbm
+shield_purge_SOURCES = shield_purge.c pam_shield_lib.c
+shield_purge_CFLAGS = $(AM_CFLAGS)
+
+install-exec-hook:
+	rm -f "${DESTDIR}${moduledir}/pam_shield.la";
+	${INSTALL} -dm0755 "${DESTDIR}${sbindir}";
+#	ln -s shield-trigger.sh "${DESTDIR}${sbindir}/pam_shield-trigger";
--- /dev/null
+++ b/autogen.sh
@@ -0,0 +1,6 @@
+#!/bin/sh
+
+export AUTOMAKE=/usr/bin/automake-1.11
+export ACLOCAL=/usr/bin/aclocal-1.11
+
+exec autoreconf -fi;
--- /dev/null
+++ b/configure.ac
@@ -0,0 +1,38 @@
+
+AC_INIT(pam_shield, 0.9.2)
+AC_CONFIG_HEADERS(config.h)
+AM_INIT_AUTOMAKE
+
+AC_PROG_CC
+AC_PROG_INSTALL
+AC_DISABLE_STATIC
+AM_PROG_LIBTOOL
+AC_PROG_LN_S
+
+AC_ARG_WITH([slibdir], AS_HELP_STRING([--with-slibdir=PATH],
+	[Path to the super lib directory [[/lib]]]),
+	[slibdir="$withval"], [slibdir="/lib"])
+AC_SUBST(slibdir)
+
+AC_CHECK_HEADER(gdbm.h, [],
+	[AC_MSG_ERROR([You need to have gdbm-devel installed])])
+AC_CHECK_HEADER(security/pam_modules.h,[have_pamheader="yes"],)
+# Mac OS X 10.3 puts PAM headers in /usr/include/pam.
+AC_CHECK_HEADER(pam/pam_modules.h,[have_pamheader="yes"],)
+if test x"$have_pamheader" != x"yes"; then
+	AC_MSG_ERROR(You are missing PAM headers)
+fi
+
+case "$host" in
+(*-*-linux*)
+	PAM_MODDIR="\$(slibdir)/security";
+	;;
+(*-*-darwin*)
+	PAM_MODDIR="/usr/lib/pam";
+	;;
+(*)
+	PAM_MODDIR="/usr/lib";
+	;;
+esac;
+AC_SUBST(PAM_MODDIR)
+AC_OUTPUT(Makefile)
