Index: pam-shield-0.9.2/pam_shield.c
===================================================================
--- pam-shield-0.9.2.orig/pam_shield.c	2007-12-05 21:00:22.000000000 +0100
+++ pam-shield-0.9.2/pam_shield.c	2007-12-05 21:00:54.000000000 +0100
@@ -164,6 +164,7 @@
 /*
 	if rhost is completely numeric, then it has no DNS entry
 */
+	if(rhost != NULL) {
 	if (strspn(rhost, "0123456789.") == strlen(rhost)
 		|| strspn(rhost, "0123456789:abcdefABCDEF") == strlen(rhost)) {
 		if (options & OPT_MISSING_DNS)
@@ -182,7 +183,7 @@
 			return PAM_IGNORE;
 		}
 	}
-
+	}
 /* if not blocking all and the user is known, let go */
 	if (!(options & OPT_BLOCK_ALL) && user != NULL && (pwd = getpwnam(user)) != NULL) {
 		logmsg(LOG_DEBUG, "ignoring known user %s", user);
