Index: pam-shield-0.9.2/shield-trigger-iptables.sh
===================================================================
--- pam-shield-0.9.2.orig/shield-trigger-iptables.sh	2007-12-05 21:15:23.000000000 +0100
+++ pam-shield-0.9.2/shield-trigger-iptables.sh	2007-12-08 21:16:36.000000000 +0100
@@ -43,7 +43,7 @@
 #	* put in the correct port number (22 is ssh)
 #	* add additional rules for additional services as needed
 #
-	"$IPT" "$1" INPUT -i eth0 -p tcp -s "$2" --destination-port 22 -j pam_shield
+	"$IPT" "$1" INPUT -p tcp -s "$2" --destination-port 22 -j pam_shield
 
 #	mail -s "[security] pam_shield blocked $2" root <<EOF
 #Another monkey kept off our backs ...
